javalang/parser.py

```diff
@@ -1805,6 +1836,27 @@ def parse_expression_3(self):
         return primary
 
     @parse_debug
+    def parse_lambda_expression(self):
+        lambda_expr = None
+        parameters = None
+        with self.tokens:
+            if self.would_accept('('):
```

Comment:
I would do something like,

```python
if self.would_accept('(', Identifier, ','):
  ...
else:
  parameters = self.parse_formal_parameters()
  ...
```
to avoid the extra nesting.

Also, you can rely on `parse_formal_parameters` to raise a `JavaSyntaxError` if there is no opening paren so its not necessary to handle this case explicitly by raising your own exception.